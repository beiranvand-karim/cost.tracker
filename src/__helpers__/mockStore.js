import configureStore from "redux-mock-store";
import costItems from "./costItems";

const mockStore = configureStore();

export default mockStore({
  costItems: {
    costItems: costItems
  },
  sourceCostItems: {
    costItems: costItems
  }
});
