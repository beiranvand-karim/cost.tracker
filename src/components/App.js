import DropdownContainer from "./CostItemCategoriesDropdown";
import CostItemsTotalSum from "./CostItemsTotalSum";
import CostItemsTable from "./CostItemsTable";
import ModalContainer from "./ModalContainer";
import { Grid, Row } from "../styles/styles";
import React, { Component } from "react";
import { connect } from "react-redux";
import { Button } from "reactstrap";
import Shield from "./Shield";

export class App extends Component {
  state = {
    modal: false
  };

  toggle = () =>
    this.setState(prevState => ({
      modal: !prevState.modal
    }));

  render() {
    const { costItems } = this.props;
    return (
      <Grid>
        <Row />
        <Row>
          <Button color="primary" onClick={this.toggle}>
            add a cost
          </Button>
        </Row>
        <Row>
          <Shield>
            <CostItemsTotalSum costItems={costItems} />
          </Shield>
        </Row>
        <Row>
          <Shield>
            <DropdownContainer />
          </Shield>
          <Shield>
            <ModalContainer toggle={this.toggle} modal={this.state.modal} />
          </Shield>
        </Row>
        <Row>
          <Shield>
            <CostItemsTable costItems={costItems} />
          </Shield>
        </Row>
      </Grid>
    );
  }
}

const mapStateToProps = state => ({
  costItems: state.costItems.costItems
});
export default connect(mapStateToProps)(App);
