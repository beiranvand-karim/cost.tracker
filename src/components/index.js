export { App } from "./App";
export { CostItemCategoriesDropdown } from "./CostItemCategoriesDropdown";
export { CostItemForm } from "./CostItemForm";
export { CostItemsTable } from "./CostItemsTable";
export { CostItemsTotalSum } from "./CostItemsTotalSum";
export { ModalContainer } from "./ModalContainer";
