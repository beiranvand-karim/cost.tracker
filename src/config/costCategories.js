export const costCategories = {
  FOOD: 0,
  CLOTHING: 1,
  RENT: 2,
  BILLS: 3
};
