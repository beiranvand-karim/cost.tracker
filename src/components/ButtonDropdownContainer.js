import React from 'react';
import {categories} from '../config/categories';
import {useDispatch, useSelector} from 'react-redux';
import {addFilteredActions} from '../actions/filteredCostActions';
import NO_FILTER from "../constants/noFilter";

export default function ButtonDropdownButtonContainer() {

   const dispatch = useDispatch();
   const originalCost = useSelector(state => state.originalCosts);
   const {costs} = originalCost;

   const select = (event) => {
      const {value} = event.target;
      if (NO_FILTER === value) {
         dispatch(addFilteredActions(costs));
         return;
      }
      const filtered = costs.filter(cost => cost.category === value);
      dispatch(addFilteredActions(filtered));
   };

   const keys = Object.keys(categories);
   keys.unshift(NO_FILTER);

   return (
      <div>
         <select onChange={select}>
            {
               keys.map((category, index) => {
                  return <option key={index} value={category}>{category}</option>
               })
            }
         </select>
      </div>
   )
}
