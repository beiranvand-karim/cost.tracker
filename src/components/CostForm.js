import {Form, FormGroup, Input, Label} from 'reactstrap';
import {categories} from '../config/categories';
import React from 'react';

export default function CostForm({cost, setCost, category, setCategory, comment, setComment}) {
   return (
      <Form>
         <FormGroup>
            <Label for="exampleEmail">Cost</Label>
            <Input
               type="email"
               name="email"
               id="exampleEmail"
               placeholder="cost here..."
               value={cost}
               onChange={(e) => setCost(e.target.value)}
            />
         </FormGroup>
         <FormGroup>
            <Label for="exampleSelect">Select</Label>
            <Input
               type="select"
               name="select"
               id="exampleSelect"
               value={category}
               onChange={(e) => setCategory(e.target.value)}
            >
               {
                  Object.keys(categories).map((category, index) => <option key={index}>{category}</option>)
               }
            </Input>
         </FormGroup>
         <FormGroup>
            <Label for="examplePassword">Comment</Label>
            <Input
               type="input"
               name="comment"
               id="comment"
               value={comment}
               onChange={(e) => setComment(e.target.value)}
               placeholder="comment here..."
            />
         </FormGroup>
      </Form>
   )
}
