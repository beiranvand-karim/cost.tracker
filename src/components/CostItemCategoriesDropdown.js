import { addFilteredCostItemsAction } from "../actions/costItemsActions";
import { costCategories } from "../config/costCategories";
import { NO_FILTER } from "../constants/constants";
import React, { Component } from "react";
import { connect } from "react-redux";

const costCategoriesKeys = Object.keys(costCategories);

export class CostItemCategoriesDropdown extends Component {
  handleChange = event => {
    const { value } = event.target;
    if (NO_FILTER === value) {
      this.props.addFilteredCostItemsAction(this.props.sourceCostItems);
      return;
    }
    const CostItems = this.props.sourceCostItems.filter(
      costItem => costItem.category === value
    );
    this.props.addFilteredCostItemsAction(CostItems);
  };

  render() {
    costCategoriesKeys.unshift(NO_FILTER);
    return (
      <div>
        <select onChange={this.handleChange}>
          {costCategoriesKeys.map((costCategory, index) => {
            return (
              <option key={index} value={costCategory}>
                {costCategory}
              </option>
            );
          })}
        </select>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addFilteredCostItemsAction: costItems =>
    dispatch(addFilteredCostItemsAction(costItems))
});
const mapStateToProps = state => ({
  sourceCostItems: state.sourceCostItems.costItems
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CostItemCategoriesDropdown);
