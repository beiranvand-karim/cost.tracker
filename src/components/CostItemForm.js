import { addSourceCostItemAction } from "../actions/sourceCostItemsActions";
import { Button, Form, FormGroup, Input, Label } from "reactstrap";
import { addCostItemAction } from "../actions/costItemsActions";
import { costCategories } from "../config/costCategories";
import React, { Component } from "react";
import { connect } from "react-redux";

const firstCategory = () => Object.keys(costCategories)[0];

export class CostItemForm extends Component {
  state = {
    cost: "",
    category: firstCategory(),
    comment: ""
  };

  handleSubmit = event => {
    event.preventDefault();
    const { category, cost, comment } = this.state;
    this.setState({ cost });
    this.setState({ comment });
    const tempCostItem = { category, cost, comment };
    this.props.addSourceCostItemAction(tempCostItem);
    this.props.addCostItemAction(tempCostItem);
    this.clearForm();
    this.props.toggle();
  };

  clearForm = () => {
    this.setState({ cost: "" });
    this.setState({ category: firstCategory() });
    this.setState({ comment: "" });
  };

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };
  render() {
    const { cost, category, comment } = this.state;
    return (
      <Form onSubmit={this.handleSubmit}>
        <FormGroup>
          <Label for="cost">Cost</Label>
          <Input
            type="input"
            name="cost"
            id="cost"
            placeholder="cost here..."
            value={cost}
            onChange={this.handleChange}
          />
        </FormGroup>
        <FormGroup>
          <Label for="category">Category</Label>
          <Input
            type="select"
            name="category"
            id="category"
            value={category}
            onChange={this.handleChange}
          >
            {Object.keys(costCategories).map((category, index) => (
              <option key={index}>{category}</option>
            ))}
          </Input>
        </FormGroup>
        <FormGroup>
          <Label for="comment">Comment</Label>
          <Input
            type="input"
            name="comment"
            id="comment"
            value={comment}
            onChange={this.handleChange}
            placeholder="comment here..."
          />
        </FormGroup>
        <Button color="primary" type="submit">
          add cost
        </Button>{" "}
        <Button color="secondary" onClick={this.props.toggle}>
          Cancel
        </Button>
      </Form>
    );
  }
}
const mapDispatchToProps = dispatch => ({
  addSourceCostItemAction: cost => dispatch(addSourceCostItemAction(cost)),
  addCostItemAction: cost => dispatch(addCostItemAction(cost))
});
export default connect(
  null,
  mapDispatchToProps
)(CostItemForm);
