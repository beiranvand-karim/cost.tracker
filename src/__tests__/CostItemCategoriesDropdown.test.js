/* global expect, describe, it, beforeEach */
import { CostItemCategoriesDropdown } from "../components/CostItemCategoriesDropdown";
import { costCategories } from "../config/costCategories";
import costItems from "../__helpers__/costItems";
import mockStore from "../__helpers__/mockStore";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { shallow } from "enzyme";
import React from "react";

describe("<CostItemCategoriesDropdown/>", () => {
  let store;
  const addFilteredCostItemActionMock = jest.fn();

  beforeEach(() => {
    store = mockStore;
  });

  it("should render without crashing", () => {
    shallow(<CostItemCategoriesDropdown />);
  });

  /**
   * this unit should be improved to cover the whole  if-else scenario
   */
  it("should push filtered cost items", () => {
    const wrapper = shallow(
      <CostItemCategoriesDropdown
        sourceCostItems={costItems}
        addFilteredCostItemsAction={addFilteredCostItemActionMock}
      />
    );

    wrapper.find("select").simulate("change", {
      target: { value: Object.keys(costCategories)[0] }
    });
    expect(addFilteredCostItemActionMock).toHaveBeenCalled();
  });

  it("should matches the snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <CostItemCategoriesDropdown />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
