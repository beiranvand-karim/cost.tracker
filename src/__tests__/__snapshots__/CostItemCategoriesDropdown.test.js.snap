// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CostItemCategoriesDropdown/> should matches the snapshot 1`] = `
<div>
  <select
    onChange={[Function]}
  >
    <option
      value="NO_FILTER"
    >
      NO_FILTER
    </option>
    <option
      value="NO_FILTER"
    >
      NO_FILTER
    </option>
    <option
      value="NO_FILTER"
    >
      NO_FILTER
    </option>
    <option
      value="FOOD"
    >
      FOOD
    </option>
    <option
      value="CLOTHING"
    >
      CLOTHING
    </option>
    <option
      value="RENT"
    >
      RENT
    </option>
    <option
      value="BILLS"
    >
      BILLS
    </option>
  </select>
</div>
`;
